lib/concurrent/atomic_reference/direct_update.rb

```diff
@@ -27,9 +27,28 @@ def update
     # @!macro [attach] atomic_reference_method_try_update
     #
     # Pass the current value to the given block, replacing it
+    # with the block's result. Return nil if the update fails.
+    #
+    # @yield [Object] Calculate a new value for the atomic reference using
+    #   given (old) value
+    # @yieldparam [Object] old_value the starting value of the atomic reference
+    #
+    # @return [Object] the new value, or nil if update failed
+    def try_update
+      old_value = get
+      new_value = yield old_value
+
+      return unless compare_and_set old_value, new_value
+
+      new_value
+    end
```

Comment:
We should be somewhat consistent with the similar function in `AtomicMarkableReference`. In that class we return an immutable array with `[new_val, new_mark]`. We could have both methods return a list with true/false as the first element. Having worked in Erlang for a few years I'm very fond of this pattern. It's very common in Erlang to return a tuple of `{ok}` or `{error message}` from functions.

If we want to go in this direction we should consider using either `ImmutableStruct` or `Maybe`. I added the latter specifically for situations like this. Maybe (aka Option) is a common pattern in functional [languages](https://hackage.haskell.org/package/base-4.2.0.1/docs/Data-Maybe.html) and [libraries](https://functionaljava.googlecode.com/svn/artifacts/3.0/javadoc/fj/data/Option.html) that I'm very fond of but that I haven't seen much in Ruby.